### ISSUE TITLE
Do we have plans for data packing?

### ISSUE THREAD

**Issue by YixinSong-e** [volcengine/verl#53]
Thank you for creating this excellent project. The design is elegant and the implementation is very efficient.

I'm wondering if there are plans to support sequence packing? 
OpenRLHF has demonstrated significant speedup after supporting sequence packing. Given VERL's already impressive performance, I believe adding packing support could bring even greater efficiency gains.

**Comment by PeterSH6** [comment]
Hi @YixinSong-e, do you mean data packing in the FSDP backend?
We already implemented data packing (i.e. remove padding?) in the Megatron-LM backend.
For the FSDP backend, we do have some monkey-patch to implement it upon transformers but it's not general for all models in transformers. So we haven't released them yet.
And we also noticed that only some important models (e.g., llama, qwen...) support data packing naturally in transformers.

We're thinking about some elegant way to support data packing for all models in our fsdp workers. Do you have any suggestions or references?

**Comment by SalmanMohammadi** [comment]
Hey @PeterSH6! 

+1 for such an awesome project. My favourite in the new wave of open-source RLHF frameworks.

> For the FSDP backend, we do have some monkey-patch to implement it upon transformers but it's not general for all models in transformers. So we haven't released them yet.

I'm curious what kind of monkey patch you mean here? I'd be interested in contributing if you'd like some help, I'm working on something similar in another repo.

**Comment by fzyzcjy** [comment]
+1 This seems to be widely used and speed up training, and I am happy to PR as well.

**Comment by PeterSH6** [comment]
Hi @SalmanMohammadi and @fzyzcjy , sorry for the late response.

The monkey patch we apply in transformers==4.44 is something like that:

```python
def flash_attn2_rmpad_forward(
        self: LlamaFlashAttention2,
        hidden_states: torch.Tensor,
        attention_mask: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.LongTensor] = None,
        past_key_value: Optional[Cache] = None,
        output_attentions: bool = False,
        use_cache: bool = False,
        cache_position: Optional[torch.LongTensor] = None,
        **kwargs  # for compatibility
):
    hidden_states = hidden_states.squeeze(0)  # (total_nnz, hidden_size)
    # this matches with hidden_states
    indices, cu_seqlens, max_seqlen_in_batch = _get_unpad_data(attention_mask)

    hidden_states = pad_input(hidden_states,
                              indices=indices,
                              batch=attention_mask.shape[0],
                              seqlen=attention_mask.shape[1])

    position_ids = compute_position_id_with_mask(attention_mask)

    attn_output, _, _ = llama_flash_attn2_orig_forward(self,
                                                       hidden_states=hidden_states,
                                                       attention_mask=attention_mask,
                                                       position_ids=position_ids,
                                                       past_key_value=past_key_value,
                                                       output_attentions=output_attentions,
                                                       use_cache=use_cache,
                                                       cache_position=cache_position,
                                                       **kwargs)

    attn_output = index_first_axis(rearrange(attn_output, "b s ... -> (b s) ..."), indices)

    return attn_output, None, None

# Monkey patches the forward function.
from transformers.models.llama.modeling_llama import LlamaFlashAttention2
LlamaFlashAttention2.forward = flash_attn2_rmpad_forward
```
With this patch, we also need to change the input to the actor model to `input_ids =  unpad_input(input_ids, attention_mask)`. Therefore, this method is hard to generalize to all models in transformers.

We also noticed that in transformers>4.45, it naturally support data packing in many models.
We really need some help to support data packing in the latest transformers. 
I've created a Slack channel for this [feature](https://verlgroup.slack.com/archives/C086K823VPH) and invited you. We can discuss the data-packing features there!

**Comment by fzyzcjy** [comment]
Btw, in SFT stage, I personally use https://huggingface.co/blog/packing-with-FA2 (e.g. trl.SFTTrainer). That one seems to work well as long as FA2 exists for a model and does not need patching. Have not checked into details though...

**Comment by PeterSH6** [comment]
Yeah, we implement this patching in older transformers version (<=4.44). I believe it's not necessary now.

**Comment by fzyzcjy** [comment]
I am thinking about another kind of data packing, which I used during SFT and seem to work well, thus propose here for discussions and happy to PR.

Suppose we have 8 sequences to be learnt by policy/critic of length 100,100,400,100,100,100,500,300, and max seq len is 500. Suppose our cuda memory only allows to process 1000 tokens in a batch.

The packing mentioned above will have to use bs=2, because we have to deal with the worst case that everything is on the max seq len. Then we will need 4 batches in this example.

The proposal is that, we can put as many sequences as possible into a batch, until it exceeds the max seq len. For example, here we can put into two batches: [100,100,400,100,100,100], [500,300]. Then we will only do 2 batches.

This is also similar to the "packing" commonly used in pretraining with some differences. In pretraining, it seems we often concat documents, and cut off whenever it reaches 2048 tokens. This is not very ideal for PPO especially for math, because it seems not very meaningful to learn "second half of a math example" without knowing the first half.
